content/blog/haskell-base-proposal.md

```diff
@@ -0,0 +1,99 @@
+* Choosing between an unboxed and storable vector is fairly tricky for newcomers, and creating appropriate instances of `Unbox` and `Storable` is far from trivial.
+
+Compare this to most other languages, where there's a canonical array, vector, or similar type which is basically used for everything. That's the story I _wish_ we had in Haskell. The question is: can we get there?
+
+## To split or not to split?
+
+The first thing that comes up in these discussions is a fairly large question about splitting base. This comes down to the idea that, if we split `base` into multiple smaller packages, everything will get better. I'm going to challenge this in two different ways:
+
+1. It's an orthogonal question. It _seems_ related because we can iterate faster on these changes if the changes can exist outside of the GHC release cycle. But I think we have a much better approach, which I'll discuss below.
+2. As I've mentioned elsewhere (not sure where), there's a false sense of immutability about `base`. People are afraid to change it due to breaking backwards compatibility. I'm in favor of backwards compatibility, and like that approach overall. However, GHC itself breaks things so regularly that it's almost irrelevant if `base` has breakage as well. I'd love to see that fixed, but I don't see it as a blocker.
+
+Given that the "split base" concept has been around for about as long as I've been using Haskell and still hasn't happened, I'm not going to hold my breath on it for this proposal.
+
+## Unifying ByteString, Text, and Vector
+
+Going back to that list of types above, I'm going to ask everyone to _ignore_ lazy lists and `Seq` as, currently, unrelated data types. They represent non-memory-packed data. Similarly, I want to ignore the lazy variants of `ByteString` and `Text` too. I think we should stop using them, but that's a discussion for another day. Instead, I want to focus in on `Vector` and its variants, `ByteString`, `Text`, and `Builder`. I also want to focus on the mutable variants and the `Unbox` and `Storable` typeclasses.
+
+First, it's long been recognized that `ByteString` is, essentially, synonymous with a storable `Vector` of `Word8`s. I say essentially, since there's a slight optimization in `Vector` to avoid holding onto an offset value. But the two are isomorphic. Representing `ByteString` as `Data.Vector.Storable.Vector Word8` is doable.
+
+Similarly, `text` implements its own unpacked array implementation. That could be replaced with `vector`'s unboxed array.
+
+So on our first step in this journey, my recommendation is simple: let's unify these types! For ease of transition, we can think of `ByteString` and `Text` as newtype wrappers for now. But as you'll see in the continuing discussion, I really have different plans.
+
+## Unify unboxed and storable vectors
+
+Credit where credit is due: Vincent Hanquez figured this out a while ago with his `foundation` package.
```

Comment:
I'd like to disagree that he figured it all out with regards to memory and arrays:

* The approach with pinned vs unpinned in `memory` is pretty dumb if you ask me, all it does is copies the memory from unpinned to pinned memory when pinned is needed (called trampoline). 
* `PrimTy` and `PrimMonad` are not that much better than `Prim` and `PrimMonad` classes from `primitive` package.
* It suffers from many the same problems that bytestring does, because quite often it expects pinned memory due to it mostly operating on Ptr
* the whole approach of forgetting about `ST` for the most part and sticking everything in `IO` is a bad decision IMHO.

There are a few good and novel ideas in `memory`+`basement`+`foundation` packages, but it is far from being better than `base`, `bytestring`, `text`, `primitive`, `vector`, etc.

Reply:
I'll stand by the comment. I'm not saying that Vincent fixed everything. I am saying that he was the first person I'm aware of who identified storable vs unboxed as a major issue in Haskell library design, and attempted to unify them.